### ISSUE TITLE
Concave Mesh (concavemesh.obj) collision problem

### ISSUE THREAD

**Issue by goksanisil23** [DanielChappuis/reactphysics3d#206]
Hi,

Firstly thank you for creating such a clean physics engine and providing sample scenes as well.

I was experimenting with using the concavemesh.obj as a static mesh as a terrain. However, I realized that at certain regions of the mesh, other dynamic objects just pass through, and in some other regions, dynamic objects get half-stuck (partially penetrate into the ground). 

You can find the phenomenon I'm talking about in the video below:

https://drive.google.com/file/d/1yAhydfGnBEB-pBhHdkzF8Zb9EZttM3YF/view?usp=sharing

I haven't seen this behavior with the city.obj concave mesh though.

Thanks in advance

**Comment by DanielChappuis** [comment]
Thanks a lot for your feedback.

I suspect that most of the issues might be because the velocities of the objects are quite large and ReactPhysics3D currently uses a discrete collision detection (not a continuous collision detection) and therefore what you observe might be a 'tunneling effect'. Try to reduce the simulation time step or make sure that the objects collide with the concave mesh with a lower velocity (by dropping them from a lower position for instance) and check if it is better.

**Comment by goksanisil23** [comment]
Hi,

Thank you for the insight.

I've tried with 0.005sec step instead of the default 0.0167 sec, as well as increasing the velocity and position solver iterations to 10x of their default values. Also, lowered the starting elevation of the falling objects. 

I believe I'm still getting similar behavior, you can see the video in the link:

https://drive.google.com/file/d/1GJVnr9OBNUXetJFIF5XuDmm-79PO2MVB/view?usp=sharing

**Comment by goksanisil23** [comment]
Similarly, I have applied the above suggestions (increasing solver resolution, decreasing impact velocity) on a different heightfield where I've also seen the tunneling effect. In that case, sufficiently increasing solver or decreasing velocity both helped with the mitigation of tunneling effect.

I believe this means that, it's a bit of tuning problem which is dependent on the static mesh characteristics.

**Comment by goksanisil23** [comment]
I wanted to give one more input.

I've generated a rectangular environment with walls in Blender and using it as a static concave mesh collider.

In the 1st video below, I'm using it without any scaling.
In the 2nd video below, I'm using it with 5x scaling on both mScalingMatrix and createConcaveMeshShape.

https://drive.google.com/file/d/1iEPgjgTFMaxUaRcCmkl0gmAoDvs1V0xa/view?usp=sharing
https://drive.google.com/file/d/1O6I-M_va25x7RdkJYG3xOzuih1bX7EBk/view?usp=sharing

And here's the code snipped where I handle the scaling (in ConcaveMesh.cpp):

`// gisil
    // Compute the scaling matrix
    mScalingMatrix = openglframework::Matrix4(5.0, 0, 0, 0,
                                              0, 5.0, 0, 0,
                                              0, 0, 5.0, 0,
                                              0, 0, 0, 1);    

    mPhysicsTriangleMesh = mPhysicsCommon.createTriangleMesh();

    // For each subpart of the mesh
    for (unsigned int i=0; i<getNbParts(); i++) {

        // Vertex and Indices array for the triangle mesh (data in shared and not copied)
        rp3d::TriangleVertexArray* vertexArray =
                new rp3d::TriangleVertexArray(getNbVertices(), &(mVertices[0]), sizeof(openglframework::Vector3),
                                              getNbFaces(i), &(mIndices[i][0]), 3 * sizeof(int),
                                              rp3d::TriangleVertexArray::VertexDataType::VERTEX_FLOAT_TYPE,
                                              rp3d::TriangleVertexArray::IndexDataType::INDEX_INTEGER_TYPE);

        // Add the triangle vertex array of the subpart to the triangle mesh
        mPhysicsTriangleMesh->addSubpart(vertexArray);
    }

    // Create the collision shape for the rigid body (convex mesh shape) and
    // do not forget to delete it at the end
    // mConcaveShape = mPhysicsCommon.createConcaveMeshShape(mPhysicsTriangleMesh);
    mConcaveShape = mPhysicsCommon.createConcaveMeshShape(mPhysicsTriangleMesh, rp3d::Vector3(5.0f, 5.0f, 5.0f)); // gisil`

**Comment by DanielChappuis** [comment]
That's interesting. Thanks a lot for this test case. 
I will try to investigate this when I have some time.

**Comment by dukeNashor** [comment]
I am in the same situation when playing with the testbed.
Loaded an OpenGL teapot mesh as a ConcaveMesh object, applied gravity, and the teapot just went through the terrain.

**Comment by DanielChappuis** [comment]
> I am in the same situation when playing with the testbed.
> Loaded an OpenGL teapot mesh as a ConcaveMesh object, applied gravity, and the teapot just went through the terrain.

Hello. This doesn't seem to be the same issue. Your are trying to collide a concave mesh (ConcaveMeshShape) against another concave mesh (HeightFieldShape). This is not supported by ReactPhysics3D (it would also be very expensive). Therefore, in your case it is expected that the two concave shapes do not collide.

**Comment by dukeNashor** [comment]
> > I am in the same situation when playing with the testbed.
> > Loaded an OpenGL teapot mesh as a ConcaveMesh object, applied gravity, and the teapot just went through the terrain.
> 
> Hello. This doesn't seem to be the same issue. Your are trying to collide a concave mesh (ConcaveMeshShape) against another concave mesh (HeightFieldShape). This is not supported by ReactPhysics3D (it would also be very expensive). Therefore, in your case it is expected that the two concave shapes do not collide.

Thanks for your reply.

**Comment by denisbogol** [comment]
@DanielChappuis, hi!

I have faced with the same issue while applying scaling to concave meshes. The problem is that the AABB for such shapes is created from the root node of DynamicAABB tree that does not account for scaling. I tried to fix in the following way:

RP3D_FORCE_INLINE void ConcaveMeshShape::getLocalBounds(Vector3& min, Vector3& max) const {

    // Get the AABB of the whole tree
    AABB treeAABB = mDynamicAABBTree.getRootAABB();

    min = treeAABB.getMin() * **mScale**;
    max = treeAABB.getMax() * **mScale**;
}
